<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('/datasets/insurance.csv')

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [4]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [5]:
data['Страховые выплаты'].value_counts()/data.shape[0]*100

0    88.72
1     8.46
2     2.30
3     0.36
4     0.14
5     0.02
Name: Страховые выплаты, dtype: float64

### Вывод
1. Пропусков нет.
2. В признаках: "Пол", "Возраст", "Члены семьи", "Страховые выплаты" нет аномальных значений. Так, признак "Пол" имеет значения 0, 1. "Возраст" от 18 до 65. "Члены семьи" от 1 до 6. 88.7% респондентов не имеют страховых выплат.Чуть более 10% имеют 1 - 2 выплаты.
3. Наименования признаков написаны с большой буквы, кириллицей. Оставим так, для большего удобства сторонних экспертов.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** Не изменится. Так как предсказания исходныи и измененных данных равны $$(a = a_n)$$

**Обоснование:** 
Обозначим новую матрицу признаков, как 

$$
X_n,  а предсказания, как
$$ 

$$
a_n,  тогда
$$

$$
X_n = XP, тогда
$$


$$
w_n = ((X_n^TX)^{-1}X_n^Ty = ((XP)^TXP)^{-1}(XP)^Ty = (P^TX^T(XP)^{-1}P^TX^Ty = P^{-1}(X^TX)^{-1}(P^T)^{-1}P^TX^Ty = P^{-1}(X^TX)^{-1}X^Ty = P^{-1}w, 
$$

так как

$$
(P^T)^{-1}P^T = 1
$$

тогда

$$
a_n = X_nw_n = XPP^{-1}w = Xw = a
$$

## Алгоритм преобразования

**Алгоритм**
1. Создадим и обучим модель линейной регрессии на исходных данных. Посмотрим значение метрики R2.
2. Защитим персональные данные клиентов, умножив матрицу исходных данных X на обратимую матрицу P.
3. Воспользуемся ранее созданной моделью на защищенных данных. Посмотрим значение метрики R2.
4. Сравним значения метрик.

**Обоснование**

Предсказяния исходных и измененных данных равны при умножении на обратимую матрицу. Такой метод шифрования допустим.

## Проверка алгоритма

Создадим и обучим модель на исходных данных

In [6]:
features = data.drop(['Страховые выплаты'], axis=1)
target = data['Страховые выплаты']

features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=.25, 
                                                                            random_state=12345)

In [7]:
print('features_train', features_train.shape)
print('target_train', target_train.shape)
print('features_valid', features_valid.shape)
print('target_valid', target_valid.shape)

features_train (3750, 4)
target_train (3750,)
features_valid (1250, 4)
target_valid (1250,)


In [8]:
model = LinearRegression()
model.fit(features_train, target_train)
predictions = model.predict(features_valid)
score = r2_score(target_valid, predictions)

print('R2 на исходных данных:', score)

R2 на исходных данных: 0.43522757127026546


Зашифруем исходные данные

In [9]:
matrix = np.random.normal(size=(features.shape[1],features.shape[1]))
inv_matrix = np.linalg.inv(matrix)
features_p = features @ inv_matrix
features_p.columns = features.columns

In [10]:
features_p.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Пол          5000 non-null   float64
 1   Возраст      5000 non-null   float64
 2   Зарплата     5000 non-null   float64
 3   Члены семьи  5000 non-null   float64
dtypes: float64(4)
memory usage: 156.4 KB


In [11]:
features_p.describe()

Пол       Возраст      Зарплата   Члены семьи
count  5000.000000   5000.000000   5000.000000   5000.000000
mean   1735.875393  -6163.348906  14103.202422  15965.492895
std     430.257934   1526.335556   3503.400642   3959.064923
min     231.591608 -12191.788096   1861.907703   2121.419240
25%    1448.681208  -7193.585988  11763.998183  13319.797693
50%    1747.129995  -6204.899478  14201.209102  16077.903347
75%    2027.250586  -5139.928163  16472.610215  18638.935944
max    3435.438144   -822.886689  27927.151305  31596.576474

In [12]:
features_p.head()

Пол      Возраст      Зарплата   Члены семьи
0  2156.742605 -7659.755349  17522.232821  19838.271820
1  1651.967688 -5872.952641  13413.190373  15199.678691
2   912.391189 -3247.164655   7409.147088   8399.409775
3  1813.955296 -6434.495574  14742.746358  16679.202389
4  1134.617177 -4033.124458   9215.014056  10438.800009

Данные зашифрованы - по ним было сложно восстановить персональную информацию.

Создадим и обучим модель на зашифрованных данных

In [13]:
features_train_p, features_valid_p, target_train, target_valid = train_test_split(features_p, target, 
                                                                                  test_size=.25, random_state=12345)

In [14]:
print('features_train_p', features_train_p.shape)
print('features_valid_p', features_valid_p.shape)
print('target_train', target_train.shape)
print('target_valid', target_valid.shape)

features_train_p (3750, 4)
features_valid_p (1250, 4)
target_train (3750,)
target_valid (1250,)


In [15]:
model.fit(features_train_p, target_train)
predictions_p = model.predict(features_valid_p)
score_p = r2_score(target_valid, predictions_p)
print('R2 на зашифрованных данных:', score_p)

R2 на зашифрованных данных: 0.4352275712702951


### Вывод
Качество линейной регрессии из sklearn не отличается до и после преобразования. Метрика R2 в обоих случаях равна 0.4352.

## Вывод

1. Были получены данные о клиентах страховой компании «Хоть потоп». Структура данных: "Пол", "Возраст", "Зарплата", "Члены семьи", "Страховые выплаты". Выбросов и аномалий в данных нет. 

2. Для защиты персональной информации, исходные данные были умножены на обратимую матрицу. Матрица данных приобрела вид, затрудняющий получение персональных данных.

3. Для проверки качества линейной регрессии, на исходных и измененных данных, были созданы модели линейной регресии. Метрика R2, отражающая качество линейной регресии, в обоих случаях, показала одинаковый результат. Качество линейней регрессии на исходных и измененных данных одинаково.

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования